src/hotspot/share/gc/shenandoah/heuristics/shenandoahAdaptiveHeuristics.cpp

```diff
@@ -110,6 +204,29 @@ bool ShenandoahAdaptiveHeuristics::should_start_gc() const {
   size_t soft_tail = max_capacity - capacity;
   available = (available > soft_tail) ? (available - soft_tail) : 0;
 
+  size_t bytes_allocated_since_gc_start = heap->bytes_allocated_since_gc_start();
+
+  // Track allocation rate even if we decide to start a cycle for other reasons.
+  _allocation_rate.sample(bytes_allocated_since_gc_start);
+  _last_trigger = OTHER;
+
+  if (is_available_below_min_threshold(capacity, available)) {
+    return true;
+  }
+
+  if (is_learning_necessary(capacity, available)) {
+    return true;
+  }
+
+  if (is_allocation_rate_too_high(capacity, available, bytes_allocated_since_gc_start)) {
+    return true;
+  }
```

Comment:
Does this split really win us anything readability-wise? I inlined the methods back, and the method is still quite readable, and does not require tracking `return`-s and pushing local variables there. I wonder if that is just a left-over from adaptive/reactive split.

Reply:
The change was certainly originally motivated by having these changes in a separate heuristic. I don't think it hurts readability, but I'll inline them back in for the sake of performance.